src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -1333,6 +1347,25 @@ public static bool UnsafeQueueUserWorkItem<TState>(Action<TState> callBack, TSta
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.callBack);
             }
 
+#if CORECLR
+            // If the callback is the runtime provided invocation of an IAsyncStateMachineBox,
+            // then we can queue the Task state directly to the ThreadPool instead of 
+            // wrapping it in a QueueUserWorkItemCallback.
+            //
+            // This occurs when user code queues its provided continuation to the ThreadPool;
+            // internally we call UnsafeQueueUserWorkItemInternal directly for Tasks.
+            if (ReferenceEquals(callBack, ThreadPoolGlobals.s_invokeAsyncStateMachineBox))
+            {
+                if (!(state is IAsyncStateMachineBox))
```

Comment:
Debug.Assert this?

Reply:
As its two fields, could be stored in a struct and be submitted during a torn write to that struct (or someone could wrongly cache the callback, but update the state). The callback itself does the type check; but here we are bypassing the callback and running directly in the threadpool when if the `state` isn't a `Task` it will do an unsafe cast: `Unsafe.As<IThreadPoolWorkItem>(workItem).Execute()`; so we can't let the dodgy state go down this path and get onto the ThreadPool.

Reply:
i.e. it wouldn't just be exception bad; but would be type system violation bad